include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -113,102 +170,124 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
         auto __reduce_pattern = unseq_backend::reduce_over_group<_ExecutionPolicy, _ReduceOp, _Tp>{__reduce_op};
 
-        const _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item);
+        _Size __size_per_work_group =
+            __iters_per_work_item1 * __work_group_size; // number of buffer elements processed within workgroup
+        _Size __n_groups = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __size_per_work_group);
+        _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item1);
 
-        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+        sycl::buffer<_Tp> __temp((sycl::range<1>(__n_groups)));
 
         sycl::event __reduce_event = __exec.queue().submit([&, __n, __n_items](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
-            auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
+            auto __temp_acc = __temp.template get_access<sycl::access_mode::write>(__cgh);
             __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size), __cgh);
-            __cgh.parallel_for<_Name...>(
-                sycl::nd_range<1>(sycl::range<1>(__work_group_size), sycl::range<1>(__work_group_size)),
+            __cgh.parallel_for<_MainName...>(
+                sycl::nd_range<1>(sycl::range<1>(__n_groups * __work_group_size), sycl::range<1>(__work_group_size)),
+                [=](sycl::nd_item<1> __item_id) {
+                    __device_reduce_kernel<_Tp>(__item_id, __n, __n_items, __transform_pattern1, __reduce_pattern,
+                                                __init, __temp_local, __temp_acc, __rngs...);
+                });
+        });
+
+        __n = __n_groups;
+
+        // Lower the work group size of the second kernel to the next power of 2 if __n < __work_group_size.
+        auto __work_group_size2 = __work_group_size;
+        if constexpr (__iters_per_work_item2 == 1)
+        {
+            if (__n < __work_group_size)
+            {
+                __work_group_size2 = __n;
+                if ((__work_group_size2 & (__work_group_size2 - 1)) != 0)
+                    __work_group_size2 = oneapi::dpl::__internal::__dpl_bit_floor(__work_group_size2) << 1;
+            }
+        }
+        __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item2);
+
+        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+
+        __reduce_event = __exec.queue().submit([&, __n, __n_items](sycl::handler& __cgh) {
+            __cgh.depends_on(__reduce_event);
+
+            auto __temp_acc = __temp.template get_access<sycl::access_mode::read>(__cgh);
+            auto __res_acc = __res.template get_access<sycl::access_mode::write>(__cgh);
+            __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size2), __cgh);
+
+            __cgh.parallel_for<_LeafName...>(
```

Comment:
We are using `_LeafName...` as the name for this kernel which includes a `work_group_size` which may deviate from the actually used one:  `work_group_size2` which can depend on `__n`.  Can this get us into trouble?  
If not, should `work_group_size` be removed from the name?

Reply:
Good point. I think we should remove `work_group_size` from the generated leaf kernel name since it has no meaning. I think we can even remove `work_group_size` from all kernel names since we only use 256.